pkg/kubelet/kubeletconfig/status/status.go

```diff
@@ -172,6 +173,11 @@ func (s *nodeConfigStatus) Sync(client clientset.Interface, nodeName string) {
 		status = status.DeepCopy()
 		status.Error = s.errorOverride
 	}
+	// Fail sync, and then retry later, in case the metrics are not created
```

Comment:
What this code is doing, at the moment, is acting as a proxy signal for kubelet initialization. I'm a bit puzzled by the current behavior. Shouldn't sync return an error if the kubelet isn't initialized anyhow?

Reply:
Thanks for the quick answer @logicalhan!

> What this code is doing, at the moment, is acting as a proxy signal for kubelet initialization. I'm a bit puzzled by the current behavior. Shouldn't sync return an error (hence retry at some future point) if the kubelet isn't initialized anyhow?

This code, the `Sync` is executed (or rather, the goroutine is dispatched) during init here: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L595, while the metrics are registered in the same function here (ofc. further down the chain): https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L720, that leads to https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1320.

I have wrapped my head around quite a few times, and from what I can see, will _only_ rerun if it results in an error. The func `Sync` will from what I understand run, but bacuase of https://github.com/kubernetes/kubernetes/pull/83184/files#diff-38e08440e8c8014419f673640b8cc03fL139, it will only run once. The only to places where we write (from what I have found) is when `Sync` fails, https://github.com/kubernetes/kubernetes/pull/83184/files#diff-38e08440e8c8014419f673640b8cc03fL156, and when creating the struct here https://github.com/kubernetes/kubernetes/pull/83184/files#diff-38e08440e8c8014419f673640b8cc03fR84.

I guess this makes sense, since kubelet restarts when a new config appears. There is still a lot of extra boiler plate, so there may have been some thoughts behind it. I also guess you @logicalhan have way better knowledge about this than me, so I may be wrong. :sweat_smile: 



Here is btw. the reason why the metrics are lost. This is a part of the new way to handle deprecation etc. of metrics, so it looks like we have to resolve this issue here, and not in the metrics system. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/gauge.go#L132. But those working on the new metrics may have some input tho.

Reply:
I see no logical issues with the code here, other than the fact that we know it will fail atleast once (aka. bad design). From what I understand, it has to be executed after the metrics have been initialized correctly here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L133

Reply:
Why can't we just move https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L593-L598 down to below where the kubelet is actually initialized?

Reply:
Metrics are lazily instantiated now to allow us to disable them. We've had issues in the past with high cardinality metrics (https://github.com/kubernetes/kubernetes/issues/73587). In order to disable metrics, we need to no-opt creation of metrics until they actually become registered, which is the current behavior. We can force metrics initialization like so (https://github.com/logicalhan/kubernetes/compare/master...kubelet-metrics-force-initialization), but I really don't want to do that. It looks to me that these tests are failing because they are literally testing that the metrics are immediately registered. That seems like a weird thing to test.